### ISSUE TITLE
Jumper seems not to work on Freeplane 1.9 ref  17

### ISSUE THREAD

**Issue by jaroet** [lilive/Freeplane-Jumper#2]
Hello,

I am rather new to Freeplane but I find the Jumper addon to be great one. I installed the release candiate of FreePlane 1.9 ref 17 and now the Jumper is not working anymore. The windows show up and I can change the search and display settings but nothing is found. The results field is empty (on activation already and that used to be all notes if I remember correctly) and searching for a value does not find anything.

I am not sure if and how I can assist you in debugging this issue. Let me know if you need some log file or something. Has anybody got this working on FreePlane 1.9?

Regards,
Jeroen

**Comment by lilive** [comment]
Hi,
Thanks for reporting. I have not tested 1.9 yet. I will let you know as soon as possible.

**Comment by lilive** [comment]
I just installed Freeplane-Setup-1.9.0-pre18 and everything seems OK for me.
Do you have the last Jumper version ?

**Comment by jaroet** [comment]
Thanks for looking into this. I was on pre17 so I installed pre18 and even removed and installed Jumper again. But when I start Jumper the window is completely blank. I do have the latest 1.0.1 installed. 

I even installed the latest release again and removed the /Users/username/.Freeplane/1.9.x/addons/jumper folder before doing a new install. Still nothing happening. 

I saw there was a lilive_jumper.json file in the 1.9.x directory so I de-installed jumper again, removed this file and the jumper directory in the addons (this is not removed by uninstalling jumper from the addons menu in Freeplane. Did a new install but still nothing.

I am not sure what I can do more.

**Comment by lilive** [comment]
Nice, these tries help me to discard some possible reasons of the trouble.
BTW, you're right, when the search field is empty, you should have 30 nodes of your map in the result list.
Have you tried with another map ?
Can you try again, and send me a copy of /Users/username/.Freeplane/1.9.x/logs/log.0, just after it ? I hope I can find some clues there.

**Comment by lilive** [comment]
Can you can also try to close FP, delete the folders compiledscripts and compiledscripts2 in .Freeplane/1.9.x, start FP and try again ? No need to uninstall and reinstall Jumper.

**Comment by jaroet** [comment]
NP ... I tried with another map and with a new build map but nothing is shown. When I look in the logs I only found the following execption ... 

STDERR: Exception in thread "AWT-EventQueue-0"
STDERR: groovy.lang.MissingMethodException: No signature of method: lilive.jumper.Candidates.fireIntervalRemoved() is applicable for argument types: (lilive.jumper.Candidates, Integer, Integer) values: [[], 0, 29]

That might give you an idea. 

The removal of the compiledscript and compiledscrits2 also did not do anything.

**Comment by lilive** [comment]
This error surprise me because the method [fireIntervalRemoved()](https://docs.oracle.com/javase/8/docs/api/javax/swing/AbstractListModel.html#fireIntervalRemoved-java.lang.Object-int-int-) is defined in the java API. Is it something special about java in your system ? What version do you use ?

You can try to install [jumper-v1.0.2.addon.mm.zip](https://github.com/lilive/Freeplane-Jumper/files/6371622/jumper-v1.0.2.addon.mm.zip) (unzip it first). I have done nothing except to recompile the .jar that comes with the add-on.

**Comment by jaroet** [comment]
I am using java8 still ...

➜  ~ java -version
openjdk version "1.8.0_242"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_242-b08)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.242-b08, mixed mode)
➜  ~

I tied wiht 1.0.2 version and even did a complete reinstall of 1.9 pre 18 release of Freeplane after removing the complete ~/.freeplane directory after deinstallation. Still nothing. 

I did a reinstall of Freeplane 1.8.11 again and created a new map. Installed Jumper 1.0.1 as 1.0.2 needs FP 1.9) and all worked fine. An empty search in Jumper showed the 5 notes my test map consisted of. 

In 1.9 not so much :-). What version of java are you running. I see in the About Freeplane that 1.8 was using 11 and 1.9 is using version 16.

**Comment by jaroet** [comment]
I am no programmer be any means but see https://docs.oracle.com/javase/7/docs/api/javax/swing/AbstractListModel.html#fireIntervalRemoved(java.lang.Object,%20int,%20int) that says the first argument should be this ? Might that be something. Now it is [] as I see in the Exception. 

Came to that by means of ... https://stackoverflow.com/questions/17411209/how-to-fix-groovy-lang-missingmethodexception-no-signature-of-method

**Comment by lilive** [comment]
Thanks for your researches !
I do use "this" when I call this method ([last lines here](https://github.com/lilive/Freeplane-Jumper/blob/master/src/main/groovy/Candidates.groovy))
I think the [] in the exception log is not a clue, but only a writing convention. But I see that you understand well some things, and that I can ask you to make more tests, isn't ?
So, I see that the groovy version have changed with FP 1.9. I have recompiled the add-on with this version. Can you try this new 
[jumper-v1.0.2.addon.mm.zip](https://github.com/lilive/Freeplane-Jumper/files/6372502/jumper-v1.0.2.addon.mm.zip) please ?
If this doesn't works I still have a couple of ideas to submit to you. I just hope this is going fast enough for you, I have not so much time these days :-)

**Comment by jaroet** [comment]
No problem ... it is just a testimony how useful I think the Jumper addon is. Using Freeplane without it is a lot less comfortable. More tests are no problem. I tried the new 1.0.2 version (first did a uninstall of everything) but still nothing. The exceptions in the logs are still the same. 

Do not worry about the speed of your reactions. I have no real issue using Freeplane without without the Jumper addon so take your time. I think communication will be rather slow because of the time differences (I guess you are from the USA?) I myself am living in the Netherlands so probably somewhere between 7 to 9 hours time difference. Looking forward to your other ideas.

**Comment by lilive** [comment]
Nice, thanks for your help.
And I'm in France :)

My best guess is that the problem come from some incompatibility at the java level.
I have an idea to test this assumption, before to try to fix it:
An add-on embed some groovy scripts. Freeplane compile them to java bytecode before it can run them.
But an add-on can also embed a pre-compiled set of scripts in a .jar file. This is what I do for Jumper. I think this .jar is not compatible with your system.
An add-on is just an easy way to install a package of groovy scripts and some other stuffs. So, you could test the scripts directly without installing the add-on, by putting them in the right place. This way, Freeplane will compile the scripts itself, and this compilation will fit your java system. Here's how to do that:

- uninstall Jumper
- close Freeplane
- restart Freeplane, this should finalize the uninstallation process
- close Freeplane again
- to be sure, check that the `fp_user_dir/addons/jumper` folder is deleted, as well as the `fp_user_dir/addons/jumper.script.xml` file
- manually delete the `fp_user_dir/lilive_jumper.json` file
- download the jumper sources: green "code" button at the top of https://github.com/lilive/Freeplane-Jumper, choose download zip
- unzip it
- copy some files it contains:

1- copy `scripts/run.groovy` into `fp_user_dir/scripts`
![2021-04-27 12h31m11s](https://user-images.githubusercontent.com/2719976/116227965-fdad0c00-a754-11eb-8b05-69715952cc0e.png)

2- create a `fp_user_dir/lib/lilive/jumper` folder
3- copy all `src/main/groovy` files into this folder
![2021-04-27 12h33m19s](https://user-images.githubusercontent.com/2719976/116228148-3351f500-a755-11eb-832e-a1f32cdba7cd.png)

- start Freeplane
- make sure you allow the scripts execution in the `Freeplane preferences > Plugin tab > Scripting section`

`Permit file/read operations`: checked
`Permit file/write operations`: checked

- make sure that Freeplane use the `fp_user_dir/lib` folder, in the same section, just a few lines below:

`Script classpath: Additional directories containing classes and/or JARs`: lib

- close the preferences and possibly restart Freeplane if you have modified the settings
- you should see the Jumper command in the menu `Edit > Find > jumper`, as usual, but the keyboard shortcut may be unset
- run this command. Freeplane will compile the scripts, this may take about 5 to 15 seconds, then the Jumper window should appear

Let me now the result, success or content of your log.0 file
Thank you.

**Comment by jaroet** [comment]
Thank you for your extensive answer and followup ... it's appreciated. 

I followed your instruction and all went rather well ... except that I do not have a Edit > Find > jumper entry but instead of jumper I get a Run option extra. See below ...
![image](https://user-images.githubusercontent.com/28713432/116234759-4a94e080-a75d-11eb-97b7-51148b158dff.png)

When I click on that the jumper screen is appearing but still empty ...
![image](https://user-images.githubusercontent.com/28713432/116235614-61880280-a75e-11eb-9066-fb3b7cf728c9.png)

- In the compiledscripts directory there is now a lilive/jamper directory with lots of .class files. That seems right to me. 
- In the compiledscripts2 directory there is a new directory with a random URI 
- In the userdir there is a lilive_jumper.json file
- The addons directory remains empty

See my log below ... 
[log.txt](https://github.com/lilive/Freeplane-Jumper/files/6384260/log.txt)

It is short but I cleared the logs just before I started Freeplane after I copied the files as mentioned in your instructions. 

I hope it helps. 

Regards,
Jeroen

**Comment by lilive** [comment]
Yes, it helps to have such a detailed feedback.

Ah ! "Run" instead of "Jumper" in the menu, this is normal, sorry.

So, you've got the same error. I'm considering to ask for help, because I can't understand nor find similar error reports in the web. To do so, it will be useful to have a minimal example of this error.
I have isolated it. Can you confirm that the following script do not works for you ?
[Test.zip](https://github.com/lilive/Freeplane-Jumper/files/6385516/Test.zip)
It is zipped because of the github attachments limitation.

- Unzip it and copy it to `fp_user_dir/scripts`
- Start (or restart) Freeplane
- Run it with the menu `Tools > Scripts > Test` entry

If it doesn't work, you should find the same
`groovy.lang.MissingMethodException: No signature of method: Test.fireIntervalAdded()`
error in the logs.

If it works you should see "Test: c d" in the status line after its execution.

**Comment by lilive** [comment]
Can you also try this other script please ?
[Test2.zip](https://github.com/lilive/Freeplane-Jumper/files/6385769/Test2.zip)
It should also print "Test: c d" in the status line.

**Comment by jaroet** [comment]
Thanks ... running the test script gives me this error

[image: image.png]

Op di 27 apr. 2021 om 17:33 schreef lilive ***@***.***>:

**Comment by jaroet** [comment]
The test2 script gives the following error ...

[image: image.png]

Here is the log file ...



Op di 27 apr. 2021 om 19:08 schreef jeroen ophetweb <
***@***.***>:

**Comment by lilive** [comment]
This one maybe ?
[Test3.zip](https://github.com/lilive/Freeplane-Jumper/files/6386731/Test3.zip)

If it don't work I think I will have enough to post my question somewhere.

**Comment by jaroet** [comment]
Test3.groovy gives me the expected result ... 

![image](https://user-images.githubusercontent.com/28713432/116294752-a16bdb80-a798-11eb-8f25-d4ddefc196a4.png)

**Comment by lilive** [comment]
OMG !
This may be the solution.
Can you try again [the manual installation of the scripts](https://github.com/lilive/Freeplane-Jumper/issues/2#issuecomment-827519969), and replace `Candidates.groovy` by this new version ?
[Candidates.zip](https://github.com/lilive/Freeplane-Jumper/files/6387085/Candidates.zip)
If you still have the previous scripts in `fp_user_dir/scripts` and `fp_user_dir/lib/lilive/jumper` , just close Freeplane, replace this file, delete `compiledscripts` and `compiledscripts2`, restart Freeplane and ~~pray~~ try jumper again.

**Comment by jaroet** [comment]
Yeah ... that did the trick. Now I have a working jumper addon again. Thank you for all your effort!
Out of curiosity, what was the OMG ! moment (in layman's terms)?

**Comment by lilive** [comment]
Nice. I will pack a new version of the add-on with this change. This is the last verification, I hope it will work too.

OMG ! for "Oh, my God !". This was... an ecstatic moment! (and this was intended to be humorous, I hope it does not hurt your feelings)

**Comment by jaroet** [comment]
Thanks ... I will check it out when it is live. 

I get the OMG phrase and the humor behind it ... no feelings hurt :-), so no problem.

I was just curious about what was the problem, I did a diff on the two different files so I already somewhat understand what the problem was.

Thanks again for all your work.

**Comment by lilive** [comment]
I realized afterward that I might misunderstood your question !

I do not understand well what is happening, so I'm not sure about how to explain it. This is perhaps the reason of the OMG reaction: when you do not understand, you invoke a power that goes beyond you :-)
This is why I probably will post something about the problem into the Freeplane forum: I'm not sure this fix is clean and reliable for the future. Also, it may be a clue that something is not 100% right about the 1.9 preview.

The best I can say to explain is:

_Fido_ is a _dog_, which is a _mammal_.
Hello Bob, here's Fido, my dog. Why do you not understand that Fido is a mammal ?

In my Test.groovy script:
_test_ is a _Candidates_, which is a _DefaultListModel_, which is an _AbstractListModel_.
Hello java, here's _test_ my _Candidates_. Why can't you do fireIntervalAdded( test ), which is supposed to works with any AbstractListModel ?

Can you try this new Test4 please ? If it works, I think it use a better and harmless way to fix the problem.
[Test4.zip](https://github.com/lilive/Freeplane-Jumper/files/6392808/Test4.zip)

... or directly test this one, following the same procedure as 4 post above
[Candidates.zip](https://github.com/lilive/Freeplane-Jumper/files/6393192/Candidates.zip)

**Comment by jaroet** [comment]
Test4 gives me again an error ... 

![image](https://user-images.githubusercontent.com/28713432/116447936-d2134a00-a858-11eb-8c51-3c4bcaf0b1af.png)

And using the Candidataes.groovy script in your last message gives me again a blank windows on activating Jumper. So the better and harmless way to fix the problem, does not work.

**Comment by lilive** [comment]
Oh :-/
Thank you. You probably have already seen that I have wrote about this issue in the forum.
Waiting for answers...

**Comment by lilive** [comment]
To explain the problem in the forum, I have simplified the test script. As a precaution, could you test it please? It should trigger the error or write "ok" in the status line. It works for me, and it should not for you.

[Test.zip](https://github.com/lilive/Freeplane-Jumper/files/6411214/Test.zip)

**Comment by jaroet** [comment]
Hi, the test.groovy script you send does indeed gives me the usual error. 

![image](https://user-images.githubusercontent.com/28713432/116811403-65929680-ab49-11eb-96fa-fc1c83cb3c64.png)

**Comment by lilive** [comment]
Thank you.
I have written a bug report https://sourceforge.net/p/freeplane/bugs/2922/

**Comment by lilive** [comment]
Can you try with FP 1.9 pre 22 ? Dimitry believes that the problem should be fixed.

**Comment by jaroet** [comment]
Hi ... I tried the FP 1.9 pre 22 and did a clean install and also installed the original 1.0.1 Jumper plugin. Everything is working like expected. So that seems to have fixed it ...

**Comment by lilive** [comment]
Nice. Thanks for your patience and for all these tries. Have you read [Dimitry's answers](https://sourceforge.net/p/freeplane/bugs/2922/#5a5c) in the bug report? The problem was not Jumper, neither Freeplane, but a issue with Groovy and Java 16.
Happy Freeplaning and Jumpering! :-D

**Comment by dpolivaev** [comment]
@jaroet @lilive vote for issue https://issues.apache.org/jira/browse/GROOVY-10070 if you want it to work with Java 16 and Java 17 LTS

**Comment by lilive** [comment]
Done. Thank you @dpolivaev .
